File: backend/src/workers/manager.ts
import { db } from '../db';
import { workers } from '../db/schema';
import { eq, sql } from 'drizzle-orm';
import { logger } from '../utils/logger';

export interface Worker {
  name: string;
  interval: number; // milliseconds
  run: () => Promise<void>;
}

export class WorkerManager {
  private workers = new Map<string, { worker: Worker; timer?: Timer; isRunning?: boolean }>();
  private isShuttingDown = false;

  constructor() {
    process.on('SIGINT', () => this.shutdown());
    process.on('SIGTERM', () => this.shutdown());
  }

  async register(worker: Worker) {
    if (this.workers.has(worker.name)) {
      throw new Error(`Worker ${worker.name} is already registered`);
    }

    this.workers.set(worker.name, { worker });
    logger.info(`Registered worker: ${worker.name}`);

    // Initialize worker in database
    try {
      await db
        .insert(workers)
        .values({
          name: worker.name,
          status: 'idle',
          runCount: 0,
          successCount: 0,
          failureCount: 0,
          isEnabled: true,
        })
        .onConflictDoNothing();
    } catch (error) {
      logger.error(`Failed to initialize worker ${worker.name} in database:`, error);
    }
  }

  async start(workerName: string) {
    const workerData = this.workers.get(workerName);
    if (!workerData) {
      throw new Error(`Worker ${workerName} not found`);
    }

    const { worker } = workerData;

    // Check if worker is enabled
    const workerRecord = await db.select().from(workers).where(eq(workers.name, workerName)).get();
    if (!workerRecord?.isEnabled) {
      logger.info(`Worker ${workerName} is disabled`);
      return;
    }

    // Run immediately (without blocking), then schedule
    this.runWorker(workerName).catch((error) => {
      logger.error(`Initial run of worker ${workerName} failed:`, error);
    });

    // Schedule periodic runs
    const timer = setInterval(async () => {
      if (!this.isShuttingDown) {
        // Run without blocking the interval
        this.runWorker(workerName).catch((error) => {
          logger.error(`Scheduled run of worker ${workerName} failed:`, error);
        });
      }
    }, worker.interval);

    this.workers.set(workerName, { worker, timer });
    logger.info(`Started worker: ${workerName} (interval: ${worker.interval}ms)`);
  }

  async stop(workerName: string) {
    const workerData = this.workers.get(workerName);
    if (!workerData || !workerData.timer) {
      return;
    }

    clearInterval(workerData.timer);
    workerData.timer = undefined;
    logger.info(`Stopped worker: ${workerName}`);
  }

  async runWorker(workerName: string) {
    const workerData = this.workers.get(workerName);
    if (!workerData) {
      throw new Error(`Worker ${workerName} not found`);
    }

    // Skip if worker is already running
    if (workerData.isRunning) {
      logger.warn(`Worker ${workerName} is already running, skipping this run`);
      return;
    }

    const { worker } = workerData;
    try {
      // Mark as running
      workerData.isRunning = true;
      this.workers.set(workerName, workerData);

      logger.info(`Running worker: ${worker.name}`);

      // Update status to running
      await db
        .update(workers)
        .set({
          status: 'running',
          lastRunAt: new Date(),
          updatedAt: new Date(),
        })
        .where(eq(workers.name, worker.name));

      // Run the worker
      await worker.run();

      // Update status to idle and increment success count
      await db
        .update(workers)
        .set({
          status: 'idle',
          runCount: sql`${workers.runCount} + 1`,
          successCount: sql`${workers.successCount} + 1`,
          lastError: null,
          nextRunAt: new Date(Date.now() + worker.interval),
          updatedAt: new Date(),
        })
        .where(eq(workers.name, worker.name));

      logger.info(`Worker ${worker.name} completed successfully`);
    } catch (error) {
      logger.error(`Worker ${worker.name} failed:`, error);

      // Update status to failed and increment failure count
      await db
        .update(workers)
        .set({
          status: 'failed',
          runCount: sql`${workers.runCount} + 1`,
          failureCount: sql`${workers.failureCount} + 1`,
          lastError: error instanceof Error ? error.message : String(error),
          nextRunAt: new Date(Date.now() + worker.interval),
          updatedAt: new Date(),
        })
        .where(eq(workers.name, worker.name));
    } finally {
      // Always mark as not running
      workerData.isRunning = false;
      this.workers.set(workerName, workerData);
    }
  }

  async startAll() {
    const enabled = process.env.WORKER_ENABLED !== 'false';
    if (!enabled) {
      logger.info('Workers are disabled by environment variable');
      return;
    }

    // Start all workers in parallel
    const startPromises = Array.from(this.workers.keys()).map(async (name) => {
      try {
        await this.start(name);
      } catch (error) {
        logger.error(`Failed to start worker ${name}:`, error);
      }
    });

    await Promise.all(startPromises);
  }

  async stopAll() {
    for (const [name] of this.workers) {
      await this.stop(name);
    }
  }

  async shutdown() {
    if (this.isShuttingDown) return;

    this.isShuttingDown = true;
    logger.info('Shutting down worker manager...');
    await this.stopAll();
    logger.info('Worker manager shutdown complete');
    process.exit(0);
  }

  async getStatus() {
    const workerRecords = await db.select().from(workers).all();
    return workerRecords.map((record) => ({
      ...record,
      isRunning:
        this.workers.has(record.name) && this.workers.get(record.name)?.timer !== undefined,
    }));
  }
}

export const workerManager = new WorkerManager();

File: backend-go/workers/manager.go
package workers

import (
	"context"
	"fmt"
	"ftoolbox/models"
	"sync"
	"time"

	"go.uber.org/zap"
	"gorm.io/gorm"
)

type WorkerManager struct {
	db          *gorm.DB
	workers     map[string]Worker
	cancelFuncs map[string]context.CancelFunc
	running     map[string]bool
	mu          sync.RWMutex
	wg          sync.WaitGroup
	enabled     bool
}

func NewWorkerManager(db *gorm.DB, enabled bool) *WorkerManager {
	return &WorkerManager{
		db:          db,
		workers:     make(map[string]Worker),
		cancelFuncs: make(map[string]context.CancelFunc),
		running:     make(map[string]bool),
		enabled:     enabled,
	}
}

// Register adds a worker to the manager
func (m *WorkerManager) Register(worker Worker) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	name := worker.Name()
	if _, exists := m.workers[name]; exists {
		return fmt.Errorf("worker %s already registered", name)
	}

	m.workers[name] = worker

	// Ensure worker exists in database
	var dbWorker models.Worker
	if err := m.db.Where("name = ?", name).First(&dbWorker).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			// Create new worker record
			dbWorker = models.Worker{
				Name:      name,
				Status:    "idle",
				IsEnabled: true,
			}
			if err := m.db.Create(&dbWorker).Error; err != nil {
				return fmt.Errorf("failed to create worker record: %w", err)
			}
		} else {
			return fmt.Errorf("failed to check worker existence: %w", err)
		}
	}

	zap.L().Info("Worker registered", zap.String("worker", name))
	return nil
}

// Start begins running a worker
func (m *WorkerManager) Start(name string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	if !m.enabled {
		zap.L().Info("Workers disabled, skipping start", zap.String("worker", name))
		return nil
	}

	worker, exists := m.workers[name]
	if !exists {
		return fmt.Errorf("worker %s not found", name)
	}

	// Check if already running
	if _, running := m.cancelFuncs[name]; running {
		zap.L().Warn("Worker already running", zap.String("worker", name))
		return fmt.Errorf("worker %s already running", name)
	}

	// Check if worker is enabled in database
	var dbWorker models.Worker
	if err := m.db.Where("name = ?", name).First(&dbWorker).Error; err != nil {
		return fmt.Errorf("failed to fetch worker status: %w", err)
	}

	if !dbWorker.IsEnabled {
		zap.L().Info("Worker is disabled", zap.String("worker", name))
		return nil
	}

	ctx, cancel := context.WithCancel(context.Background())
	m.cancelFuncs[name] = cancel

	m.wg.Add(1)
	go m.runWorker(ctx, worker)

	zap.L().Info("Worker started", zap.String("worker", name))
	return nil
}

// Stop halts a running worker
func (m *WorkerManager) Stop(name string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	cancel, exists := m.cancelFuncs[name]
	if !exists {
		return fmt.Errorf("worker %s not running", name)
	}

	cancel()
	delete(m.cancelFuncs, name)

	zap.L().Info("Worker stopped", zap.String("worker", name))
	return nil
}

// StopAll stops all running workers
func (m *WorkerManager) StopAll() {
	m.mu.Lock()
	for name, cancel := range m.cancelFuncs {
		cancel()
		delete(m.cancelFuncs, name)
		zap.L().Info("Worker stopped", zap.String("worker", name))
	}
	m.mu.Unlock()

	// Wait for all workers to finish
	m.wg.Wait()
	zap.L().Info("All workers stopped")
}

// GetStatus returns the status of all workers
func (m *WorkerManager) GetStatus() ([]models.Worker, error) {
	var workers []models.Worker
	if err := m.db.Find(&workers).Error; err != nil {
		return nil, err
	}
	return workers, nil
}

func (m *WorkerManager) runWorker(ctx context.Context, worker Worker) {
	defer m.wg.Done()

	ticker := time.NewTicker(worker.Interval())
	defer ticker.Stop()

	// Run immediately
	m.executeWorker(ctx, worker)

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			m.executeWorker(ctx, worker)
		}
	}
}

func (m *WorkerManager) executeWorker(ctx context.Context, worker Worker) {
	name := worker.Name()

	// Check if already running
	m.mu.RLock()
	if m.running[name] {
		m.mu.RUnlock()
		zap.L().Debug("Worker already running, skipping", zap.String("worker", name))
		return
	}
	m.mu.RUnlock()

	// Mark as running
	m.mu.Lock()
	m.running[name] = true
	m.mu.Unlock()

	defer func() {
		m.mu.Lock()
		m.running[name] = false
		m.mu.Unlock()
	}()

	// Update status to running
	now := time.Now()
	if err := m.db.Model(&models.Worker{}).
		Where("name = ?", name).
		Updates(map[string]interface{}{
			"status":      "running",
			"last_run_at": now,
			"updated_at":  now,
		}).Error; err != nil {
		zap.L().Error("Failed to update worker status", zap.String("worker", name), zap.Error(err))
		return
	}

	// Run the worker
	startTime := time.Now()
	err := worker.Run(ctx)
	duration := time.Since(startTime)

	// Update worker status based on result
	status := "idle"
	updates := map[string]interface{}{
		"status":      status,
		"run_count":   gorm.Expr("run_count + 1"),
		"updated_at":  time.Now(),
		"next_run_at": time.Now().Add(worker.Interval()),
	}

	if err != nil {
		status = "failed"
		updates["status"] = status
		updates["failure_count"] = gorm.Expr("failure_count + 1")
		updates["last_error"] = err.Error()

		zap.L().Error("Worker failed",
			zap.String("worker", name),
			zap.Error(err),
			zap.Duration("duration", duration))
	} else {
		updates["success_count"] = gorm.Expr("success_count + 1")
		updates["last_error"] = nil

		zap.L().Info("Worker completed",
			zap.String("worker", name),
			zap.Duration("duration", duration))
	}

	if err := m.db.Model(&models.Worker{}).
		Where("name = ?", name).
		Updates(updates).Error; err != nil {
		zap.L().Error("Failed to update worker status", zap.String("worker", name), zap.Error(err))
	}
}
